code/modules/reagents/chemistry_machinery/xenomorph_analyzer.dm

```diff
@@ -0,0 +1,155 @@
+	if(!skillcheck(user, SKILL_RESEARCH, SKILL_RESEARCH_TRAINED))
+		to_chat(user, SPAN_WARNING("You have no idea how to use this."))
+		return
+	if(istype(W, /obj/item/organ/xeno))
+		if(!do_after(user, 3 SECONDS, INTERRUPT_ALL, BUSY_ICON_GENERIC))
+			to_chat(user, SPAN_WARNING("You were interupted!"))
+			return
+		if(!user.drop_inv_item_to_loc(W, src))
+			return
+		to_chat(user, SPAN_NOTICE("You place the organ in the machine"))
+		organ = W
+		icon_state = "xeno_analyzer_organ_on"
+		caste_of_organ = organ.caste_origin
+	if(istype(W, /obj/item/clothing/accessory/health/research_plate))
+		var/obj/item/clothing/accessory/health/research_plate/plate = W
+		if(plate.recyclable_value == 0 && !plate.can_recycle(user))
+			to_chat(user, SPAN_WARNING("You cannot recycle this type of plate"))
+			return
+		if(!do_after(user, 3 SECONDS, INTERRUPT_ALL, BUSY_ICON_GENERIC))
+			to_chat(user, SPAN_WARNING("You were interupted!"))
+			return
+		to_chat(user, SPAN_NOTICE("You recycle the plate"))
+		biomass_points += plate.recyclable_value
+		qdel(W)
+
+/obj/structure/machinery/xenoanalyzer/ui_data(mob/user)
+	var/list/data = list()
+	data["points"] = biomass_points
+	data["current_clearance"] = GLOB.chemical_data.clearance_level
+	data["is_x_level"] = GLOB.chemical_data.reached_x_access // why just why
+
+	if(organ)
+		data["organ"] = TRUE
+		data["caste"] = caste_of_organ
+		data["value"] = organ.research_value
+	else
+		data["organ"] = FALSE
+	return data
+
+/obj/structure/machinery/xenoanalyzer/ui_static_data(mob/user)
+	var/list/static_data = list()
+	static_data["upgrades"] = list()
+	static_data["categories"] = list()
+	for(var/upgrade_type in subtypesof(/datum/research_upgrades))
+		var/datum/research_upgrades/upgrade = upgrade_type
+		if(upgrade.behavior == RESEARCH_UPGRADE_CATEGORY)
+			static_data["categories"] += upgrade.name
+			continue
+		if(upgrade.behavior == RESEARCH_UPGRADE_EXCLUDE_BUY)
+			continue
+		static_data["upgrades"] += list(list(
+			"name" = capitalize_first_letters(upgrade.name),
+			"desc" = upgrade.desc,
+			"vari" = upgrade.behavior,
+			"cost" = upgrade.value_upgrade,
+			"ref" = upgrade.item_reference,
+			"category" = upgrade.upgrade_type,
+			"clearance" = upgrade.clearance_req,
+		))
+	return static_data
+
+/obj/structure/machinery/xenoanalyzer/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if(.)
+		return
+
+	switch(action)
+		if("eject_organ")
+			eject_biomass()
+			. = TRUE
+
+		if("process_organ")
+			addtimer(CALLBACK(src, PROC_REF(process_organ)), 5 SECONDS)
+			icon_state = "xeno_analyzer_on_moving"
+			playsound(src.loc, 'sound/machines/blender.ogg', 25, TRUE)
+			. = TRUE
+		if("produce")
+			var/cost = text2num(params["cost"])
+			var/vari = text2num(params["varia"])
+			var/clearance_req = text2num(params["clearreq"])
+			if(cost && !busy)
+				start_print_upgrade(params["ref"], cost, usr, vari, clearance_req)
```

Comment:
okay this is my lesson about not trusting user input:

currently, anyone could come up to this ui and literally print off anything in the game. you can *never* trust that stuff you're getting back from tgui (or any ui) is safe data, as they can end up sending whatever data back to us.

currently, we take a user input for cost, vari and the *path of the item*. none of these are safe, but the latter is certainly the least safe. instead, you'll need to check that the path they're sending back to you is of the subtype you're expecting, and then get the cost and vari variable from *that* instead

Reply:
done I hope